alf/utils/dist_utils.py

```diff
@@ -361,23 +359,84 @@ def estimated_entropy(dist,
+        entropy = entropy.mean(dim=0)
+        entropy_for_gradient = entropy_for_gradient.mean(dim=0)
     return entropy, entropy_for_gradient
 
 
+def entropy_with_fallback(distributions, action_spec):
+    """Computes total entropy of nested distribution.
+    If entropy() of a distribution is not implemented, this function will
+    fallback to use sampling to calculate the entropy. It returns two values:
+    (entropy, entropy_for_gradient).
+    There are two situations:
+    * entropy() is implemented. entropy is same as entropy_for_gradient.
+    * entropy() is not implemented. We use sampling to calculate entropy. The
+        unbiased estimator for entropy is -log(p(x)). However, the gradient of
+        -log(p(x)) is not an unbiased estimator of the gradient of entropy. So
+        we also calculate a value whose gradient is an unbiased estimator of
+        the gradient of entropy. See estimated_entropy() for detail.
+    Example:
+        ent, ent_for_grad = entropy_with_fall_back(dist, action_spec)
+        alf.summary.scalar("entropy", ent)
+        ent_for_grad.backward()
+    Args:
+        distributions (nested Distribution): A possibly batched tuple of
+            distributions.
+        action_spec (nested BoundedTensorSpec): A nested tuple representing the
+            action spec.
+    Returns:
+        tuple of (entropy, entropy_for_gradient). You should use entropy in
+        situations where its value is needed, and entropy_for_gradient where
+        you need to calculate the gradient of entropy.
+    """
+
+    def _calc_outer_rank(dist: td.Distribution, action_spec):
+        if isinstance(dist, td.TransformedDistribution):
+            # TransformedDistribution does not implement the two necessary
+            # interface functions of Distribution. So we have to use the
+            # original distribution it transforms.
+            # TransformedDistribution is used by NormalProjectionNetwork with
+            # scale_distribution=True
+            dist = dist.base_dist
+        return (len(dist.batch_shape) + len(dist.event_shape) - len(
+            action_spec.shape))
+
+    def _compute_entropy(dist: td.Distribution, action_spec):
+        if isinstance(dist, td.TransformedDistribution):
+            entropy, entropy_for_gradient = estimated_entropy(dist)
+        else:
+            entropy = dist.entropy()
+            entropy_for_gradient = entropy
+
+        outer_rank = _calc_outer_rank(dist, action_spec)
+        rank = len(entropy.shape)
+        reduce_dims = list(range(outer_rank, rank))
+        entropy = torch.sum(entropy, reduce_dims)
+        entropy_for_gradient = torch.sum(entropy_for_gradient, reduce_dims)
+        return entropy, entropy_for_gradient
+
+    entropies = list(
+        map(_compute_entropy, nest.flatten(distributions),
+            nest.flatten(action_spec)))
+    entropies_for_gradient = [eg for e, eg in entropies]
+    entropies = [e for e, eg in entropies]
```

Comment:
No need to loop twice. `entropies, entropies_for_gradient = zip(*entropies)`

Reply:
Nice!